federation/pkg/federation-controller/service/endpoint_helper.go

```diff
@@ -0,0 +1,136 @@
+	// obj holds the latest service info from apiserver
+	obj, exists, err := clusterCache.endpointStore.GetByKey(key)
+	if err != nil {
+		glog.Infof("Unable to retrieve endpoint %v from store: %v", key, err)
+		clusterCache.endpointQueue.Add(key)
+		return err
+	}
+	if !exists {
+		// service absence in store means watcher caught the deletion, ensure LB info is cleaned
+		glog.Infof("endpoint has been deleted %v", key)
+		cc.processEndpointDeletion(cachedService, key, clusterName)
+	}
+	if exists {
+		endpoint, ok := obj.(*v1.Endpoints)
+		if ok {
+			glog.V(4).Infof("Found endpoint for federation service %s/%s from cluster %s", endpoint.Namespace, endpoint.Name, clusterName)
+			cc.processEndpointUpdate(cachedService, endpoint, fedClient, clusterName)
+		} else {
+			tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+			if !ok {
+				return fmt.Errorf("object contained wasn't a service or a deleted key: %+v", obj)
+			}
+			glog.Infof("Found tombstone for %v", key)
+			cc.processEndpointDeletion(cachedService, tombstone.Key, clusterName)
+		}
+	}
+
+	return nil
+}
+
+func (cc *clusterClientCache) processEndpointDeletion(cachedService *cachedService, obj interface{}, clusterName string) {
+	glog.V(4).Infof("Processing endpoint update for %s, cluster %s", obj, clusterName)
+	cachedService.rwlock.Lock()
+	defer cachedService.rwlock.Unlock()
+	_, ok := cachedService.endpointMap[clusterName]
+	if ok {
+		// endpoints lost, clean dns record
+		glog.V(4).Infof("cached endpoint was not found for %s, cluster %s, building one", obj, clusterName)
+		//TODO
+		ensureDNSRecords(clusterName, cachedService)
+		delete(cachedService.endpointMap, clusterName)
+	}
+}
+
+// Update dns info when endpoint update event received
+// We do not care about the endpoint info, what we need to make sure here is len(endpoints.subsets)>0
+func (cc *clusterClientCache) processEndpointUpdate(cachedService *cachedService, endpoint *v1.Endpoints, fedClient federation_release_1_3.Interface, clusterName string) {
+	glog.V(4).Infof("Processing endpoint update for %s/%s, cluster %s", endpoint.Namespace, endpoint.Name, clusterName)
+	cachedService.rwlock.Lock()
+	defer cachedService.rwlock.Unlock()
+	_, ok := cachedService.endpointMap[clusterName]
+	if !ok {
+		// first time get endpoints, update dns record
+		glog.V(4).Infof("cached endpoint was not found for %s/%s, cluster %s, building one", endpoint.Namespace, endpoint.Name, clusterName)
+		cachedService.endpointMap[clusterName] = len(endpoint.Subsets)
+		ensureDNSRecords(clusterName, cachedService)
+	}
+}
+// TODO
+func updateDNSInfo() {
+
+}
+
+// obj could be an *api.Service, or a DeletionFinalStateUnknown marker item.
+func (cc *clusterClientCache) enqueueEndpoint(obj interface{}, clusterName string) {
+	key, err := controller.KeyFunc(obj)
+	if err != nil {
+		glog.Errorf("Couldn't get key for object %+v: %v", obj, err)
+		return
+	}
+	cc.rwlock.Lock()
+	defer cc.rwlock.Unlock()
+	_, ok := cc.clientMap[clusterName]
+	if ok {
```

Comment:
it's kind of tricky, yet, if we lock the map here, we lose parallel, but if we do not lock the map, the cluster map can be delete when delFromClusterSet is triggered.
I remove the lock here, any more suggestion?